GDCpp/GDCpp/Runtime/PolygonCollision.cpp

```diff
@@ -129,6 +137,130 @@ CollisionResult GD_API PolygonCollisionTest(Polygon2d & p1, Polygon2d & p2)
+    //     float sqRadius = (circleX - poly.vertices[1].x)*(circleX - poly.vertices[1].x) +
+    //                      (circleY - poly.vertices[1].y)*(circleY - poly.vertices[1].y);
+    //     float dx = endX - startX;
+    //     float dy = endY - startY;
+
+    //     float a = dx*dx + dy*dy;
+    //     float b = 2*(dx*(startX - circleX) + dy*(startY - circleY));
+    //     float c = (startX - circleX)*(startX - circleX) + (startY - circleY)*(startY - circleY) - sqRadius;
+    //     float det = b*b - 4*a*c;
+
+    //     if ( a == 0 || det < 0 ) return result;
+
+    //     if ( det == 0 )
+    //     {
+    //         float t = -b/(2*a);
+    //         if ( 0 <= t && t <= 1 ) {
+    //             result.closePoint.x = startX + t*dx;
+    //             result.closePoint.y = startY + t*dy;
+    //             result.closeSqDist = t*t*a;
+    //             result.farPoint = result.closePoint;
+    //             result.farSqDist = result.closeSqDist;
+    //             result.collision = true;
+    //         }
+    //     }
+    //     else
+    //     {
+    //         float sqDet = sqrt(det);
+
+    //         float t = (-b + sqDet)/(2*a);
+    //         if ( 0 <= t && t <= 1 ) {
+    //             result.closePoint.x = startX + t*dx;
+    //             result.closePoint.y = startY + t*dy;
+    //             result.closeSqDist = t*t*a;
+    //             result.farPoint = result.closePoint;
+    //             result.farSqDist = result.closeSqDist;
+    //             result.collision = true;
+    //         }
+
+    //         t = (-b - sqDet)/(2*a);
+    //         if ( 0 <= t && t <= 1 ){
+    //             result.closePoint.x = startX + t*dx;
+    //             result.closePoint.y = startY + t*dy;
+    //             result.closeSqDist = t*t*a;
+    //             if ( !result.collision ) {
+    //                 result.farPoint = result.closePoint;
+    //                 result.farSqDist = result.closeSqDist;
+    //             }
+    //             result.collision = true;
+    //         }
+    //     }
+
+    //     return result;
+    // }
+    // else
+    // {
+        // Polygon raycasting
+        poly.ComputeEdges();
+
+        sf::Vector2f p, q, r, s;
+        float minSqDist = FLT_MAX;
+        // ray segment: p + t*r, with p = start and r = end - start
+        p.x = startX;
+        p.y = startY;
+        r.x = endX - startX;
+        r.y = endY - startY;
+
+        for( int i=0; i<poly.edges.size(); i++ )
+        {
+            // edge segment: q + u*s
+            q = poly.vertices[i];
+            s = poly.edges[i];
+
+            sf::Vector2f deltaQP = q - p;
+            float crossRS = crossProduct(r, s);
+            float t = crossProduct(deltaQP, s) / crossRS;
+            float u = crossProduct(deltaQP, r) / crossRS;
+            
+
+            if ( crossRS == 0 && crossProduct(deltaQP, r) == 0 )
+            {
+                // TODO Collinear

```

Comment:
Do you know what is still to be done to handle this case? :)

Reply:
I was thinking about it but it's a bit cumbersome, it's a super particular case (even more particular in games), but it would be nice to add it. The idea is: the ray and the edge are collinear/overlaps, so you have to calculate the closest/farthest overlapping point. The problem is that the only solution I can think so far is multiple if/else nested, to check ray direction, condition inverted and overlapping interval, so I gave up... for now :(